File: examples/c/tests/test_vector3.c
#include "test_runner.h"
#include "vector3.h"

#define EPS 1e-10

int main(void) {
    DESCRIBE("Vector3");

    IT("creates with values", {
        Vec3 v = vec3_create(1, 2, 3);
        EXPECT_NEAR(v.x, 1.0, EPS);
        EXPECT_NEAR(v.y, 2.0, EPS);
        EXPECT_NEAR(v.z, 3.0, EPS);
    });

    IT("computes length", {
        Vec3 v = vec3_create(3, 4, 0);
        EXPECT_NEAR(vec3_length(v), 5.0, EPS);
    });

    IT("normalizes", {
        Vec3 v = vec3_create(0, 0, 5);
        Vec3 n = vec3_normalized(v);
        EXPECT_NEAR(n.x, 0.0, EPS);
        EXPECT_NEAR(n.y, 0.0, EPS);
        EXPECT_NEAR(n.z, 1.0, EPS);
    });

    IT("adds", {
        Vec3 r = vec3_add(vec3_create(1, 2, 3), vec3_create(4, 5, 6));
        EXPECT_NEAR(r.x, 5.0, EPS);
        EXPECT_NEAR(r.y, 7.0, EPS);
        EXPECT_NEAR(r.z, 9.0, EPS);
    });

    IT("subtracts", {
        Vec3 r = vec3_sub(vec3_create(5, 7, 9), vec3_create(1, 2, 3));
        EXPECT_NEAR(r.x, 4.0, EPS);
        EXPECT_NEAR(r.y, 5.0, EPS);
        EXPECT_NEAR(r.z, 6.0, EPS);
    });

    IT("scales", {
        Vec3 r = vec3_scale(vec3_create(1, 2, 3), 2.0);
        EXPECT_NEAR(r.x, 2.0, EPS);
        EXPECT_NEAR(r.y, 4.0, EPS);
        EXPECT_NEAR(r.z, 6.0, EPS);
    });

    IT("computes dot product", {
        Vec3 a = vec3_create(1, 2, 3);
        Vec3 b = vec3_create(4, 5, 6);
        EXPECT_NEAR(vec3_dot(a, b), 32.0, EPS);
    });

    IT("computes cross product", {
        Vec3 a = vec3_create(1, 0, 0);
        Vec3 b = vec3_create(0, 1, 0);
        Vec3 c = vec3_cross(a, b);
        EXPECT_NEAR(c.x, 0.0, EPS);
        EXPECT_NEAR(c.y, 0.0, EPS);
        EXPECT_NEAR(c.z, 1.0, EPS);
    });

    IT("checks equality", {
        Vec3 a = vec3_create(1, 2, 3);
        Vec3 b = vec3_create(1, 2, 3);
        EXPECT_TRUE(vec3_equal(a, b));
        EXPECT_TRUE(!vec3_equal(a, vec3_create(0, 0, 0)));
    });

    SUMMARY();
}

File: examples/rust/tests/test_vector3.rs
extern crate shapes;

use shapes::vector3::Vec3;

const EPS: f64 = 1e-10;

#[test]
fn creates_with_values() {
    let v = Vec3::new(1.0, 2.0, 3.0);
    assert!((v.x - 1.0).abs() < EPS);
    assert!((v.y - 2.0).abs() < EPS);
    assert!((v.z - 3.0).abs() < EPS);
}

#[test]
fn computes_length() {
    let v = Vec3::new(3.0, 4.0, 0.0);
    assert!((v.length() - 5.0).abs() < EPS);
}

#[test]
fn normalizes() {
    let n = Vec3::new(0.0, 0.0, 5.0).normalized();
    assert!((n.x - 0.0).abs() < EPS);
    assert!((n.y - 0.0).abs() < EPS);
    assert!((n.z - 1.0).abs() < EPS);
}

#[test]
fn adds() {
    let r = Vec3::new(1.0, 2.0, 3.0) + Vec3::new(4.0, 5.0, 6.0);
    assert!((r.x - 5.0).abs() < EPS);
    assert!((r.y - 7.0).abs() < EPS);
    assert!((r.z - 9.0).abs() < EPS);
}

#[test]
fn subtracts() {
    let r = Vec3::new(5.0, 7.0, 9.0) - Vec3::new(1.0, 2.0, 3.0);
    assert!((r.x - 4.0).abs() < EPS);
    assert!((r.y - 5.0).abs() < EPS);
    assert!((r.z - 6.0).abs() < EPS);
}

#[test]
fn scales() {
    let r = Vec3::new(1.0, 2.0, 3.0) * 2.0;
    assert!((r.x - 2.0).abs() < EPS);
    assert!((r.y - 4.0).abs() < EPS);
    assert!((r.z - 6.0).abs() < EPS);
}

#[test]
fn computes_dot_product() {
    let a = Vec3::new(1.0, 2.0, 3.0);
    let b = Vec3::new(4.0, 5.0, 6.0);
    assert!((a.dot(&b) - 32.0).abs() < EPS);
}

#[test]
fn computes_cross_product() {
    let c = Vec3::new(1.0, 0.0, 0.0).cross(&Vec3::new(0.0, 1.0, 0.0));
    assert!((c.x - 0.0).abs() < EPS);
    assert!((c.y - 0.0).abs() < EPS);
    assert!((c.z - 1.0).abs() < EPS);
}

#[test]
fn checks_equality() {
    let a = Vec3::new(1.0, 2.0, 3.0);
    assert_eq!(a, Vec3::new(1.0, 2.0, 3.0));
    assert_ne!(a, Vec3::new(0.0, 0.0, 0.0));
}
